lib/tables.c

```diff
@@ -1756,16 +1783,414 @@ static int
 migration_table_load(migration_table_t *self, kastore_t *store)
 {
     read_table_col_t read_cols[] = {
-        {"migrations/left", (void **) &self->left, &self->num_rows, 0, KAS_FLOAT64},
-        {"migrations/right", (void **) &self->right, &self->num_rows, 0, KAS_FLOAT64},
-        {"migrations/node", (void **) &self->node, &self->num_rows, 0, KAS_INT32},
-        {"migrations/source", (void **) &self->source, &self->num_rows, 0, KAS_INT32},
-        {"migrations/dest", (void **) &self->dest, &self->num_rows, 0, KAS_INT32},
-        {"migrations/time", (void **) &self->time, &self->num_rows, 0, KAS_FLOAT64},
+        {"migrations/left", (void **) &self->left, &self->num_rows, 0, 1, KAS_FLOAT64},
+        {"migrations/right", (void **) &self->right, &self->num_rows, 0, 1, KAS_FLOAT64},
+        {"migrations/node", (void **) &self->node, &self->num_rows, 0, 1, KAS_INT32},
+        {"migrations/source", (void **) &self->source, &self->num_rows, 0, 1, KAS_INT32},
+        {"migrations/dest", (void **) &self->dest, &self->num_rows, 0, 1, KAS_INT32},
+        {"migrations/time", (void **) &self->time, &self->num_rows, 0, 1, KAS_FLOAT64},
     };
     return read_table_cols(store, read_cols, sizeof(read_cols) / sizeof(*read_cols));
 }
 
+/*************************
+ * individual table
+ *************************/
+
+static int
+individual_table_expand_main_columns(individual_table_t *self, table_size_t additional_rows)
+{
+    int ret = 0;
+    table_size_t increment = MSP_MAX(additional_rows, self->max_rows_increment);
+    table_size_t new_size = self->max_rows + increment;
+
+    if ((self->num_rows + additional_rows) > self->max_rows) {
+        ret = expand_column((void **) &self->flags, new_size, sizeof(uint32_t));
+        if (ret != 0) {
+            goto out;
+        }
+        ret = expand_column((void **) &self->location, new_size,
+                self->spatial_dimension * sizeof(double));
+        if (ret != 0) {
+            goto out;
+        }
+        ret = expand_column((void **) &self->metadata_offset, new_size + 1,
+                sizeof(table_size_t));
+        if (ret != 0) {
+            goto out;
+        }
+        self->max_rows = new_size;
+    }
+out:
+    return ret;
+}
+
+static int
+individual_table_expand_metadata(individual_table_t *self, table_size_t additional_length)
+{
+    int ret = 0;
+    table_size_t increment = MSP_MAX(additional_length,
+            self->max_metadata_length_increment);
+    table_size_t new_size = self->max_metadata_length + increment;
+
+    if ((self->metadata_length + additional_length) > self->max_metadata_length) {
+        ret = expand_column((void **) &self->metadata, new_size, sizeof(char *));
+        if (ret != 0) {
+            goto out;
+        }
+        self->max_metadata_length = new_size;
+    }
+out:
+    return ret;
+}
+
+int
+individual_table_set_spatial_dimension(individual_table_t *self, table_size_t spatial_dimension)
```

Comment:
This is the only function that doesn't follow the pattern of other tables. it seems necessary because of the `realloc`ing... but it might be more natural if one could just set the `spatial_dimension` directly?